### DISCUSSION TITLE
How to custom CSV Download

### DISCUSSION THREAD

**Discussion by kyowanghwe** [dentsusoken/iPLAss#1666] [Q&A]
I want to change the column name in a CSV file, but I'm not sure how to do it. Can we use SearchQueryInterrupter for this because our old system adready implement SearchQueryInterrupter for CSV Download action? 
Could you guys provide an example, tks you so much

![image](https://github.com/user-attachments/assets/290a3be0-9b12-43be-9f20-c42a261808dc)

**Comment by hkondoh** [comment]
Hi @kyowanghwe,

You cannot customize the column names of the CSV file by the `SearchQueryInterrupter`.
The column names are taken from a SearchView definition ( `Label` ) or an Entity Definition ( `Display Name` ) by default.

To customize these column names in the CSV file, you need to create a custom download action and set it in the `CSV Download Action` (or `Download Action` if you using version 4.0.0 or above) of the SearchFormView Setting.
When creating the custom download action, follow these steps:

1. Create a Command that implements both the search process for Entity data and a `ResultStreamWriter` for data output.
2. Configure the Results of the action to output the CSV file as a binary stream.

[Action Result](https://iplass.org/docs/developerguide/customizing/index.html#Action-Result)
[Result: Stream](https://iplass.org/docs/developerguide/customizing/index.html#Result-Stream)
[EntityFileDownloadSearchViewWriter.java](https://github.com/dentsusoken/iPLAss/blob/master/iplass-gem/src/main/java/org/iplass/gem/command/generic/search/EntityFileDownloadSearchViewWriter.java)

Alternatively, you can also customize the column names using the `columnNameMap` method of the `SearchFormCsvUploadInterrupter` .
However, this interrupter can only be used when downloading files in the `Output format for Uploading` mode.
[Upload Customization](https://iplass.org/docs/developerguide/genericentitymanager/entityview/index.html#GemCustomize_Upload)

It looks like your old system already has a custom download action for downloading the CSV file.
I recommend trying to port this action to the new system first.
(Please note that you may need to adjust the configuration or source code when migrating it to the new system.)
